device/rsvp/rsvpFastReroute.yaml

```diff
@@ -0,0 +1,76 @@
+components:
+  schemas:
+    Rsvp.FastReroute:
+      description: >-
+        Configuration for the optional RSVP-TE FAST_REROUTE object included in Path Messages as defined in RFC4090. 
+      type: object
+      properties:
+        setup_priority:
+          description: >-
+            Specifies the value of the Setup Priority field. This controls whether the backup LSP should pre-empt existing
+            LSP that is setup with certain Holding Priority. While setting up a backup LSP, preemption of existing LSP can happen 
+            if resource limitation is encountered (e.g bandwidth availability).
+          type: integer
+          default: 7
+          minimum: 0
+          maximum: 7
+        holding_priority:
+          description: >-
+            Specifies the value of the Holding Priority field. This controls whether a new LSP being created with certain Setup Priority
+            should pre-empt this LSP set up with this Holding Priority. While setting up a new LSP, preemption of existing LSP can happen 
+            if resource limitation is encountered (e.g bandwidth availability).            
+          type: integer
+          default: 7
+          minimum: 0
+          maximum: 7
+        hop_limit:
+          description: >-
+            Specifies the value of the Hop Limit field. This controls the maximum number of hops the LSP should traverse to reach the 
+            LSP end-point.
+          type: integer
+          default: 3
+          minimum: 0
+          maximum: 255
+        bandwidth:
+          description: >-
+            Specifies the value of the Bandwidth field as a 32-bit IEEE floating point integer, in
+            bytes per second, as desired for the LSP. 
+          type: number
+          format: float
+          default: 0
+        exclude_any:
+          description: >-
+            A 32-bit vector representing a set of attribute filters associated with a tunnel any of which renders a link
+            unacceptable. A null set (all bits set to zero) doesn't render the link unacceptable.            
+          type: string
+          format : hex
+          default: '0'                  
+        include_any:
+          description: >-
+            A 32-bit vector representing a set of attribute filters associated with a tunnel any of which renders a link
+            acceptable. A null set (all bits set to zero) automatically passes.       
+          type: string
+          format : hex
+          default: '0'     
+        include_all:
+          description: >-
+            A 32-bit vector representing a set of attribute filters associated with a tunnel all of which must be present for a
+            link to be acceptable. A null set (all bits set to zero) automatically passes.          
+          type: string
```

Comment:
Set min/max length to 0/8.

Reply:
Setting the min and max might not be enough.  What is the string specified is '0101'.  What is the value assumed for the remaining bits?  And which bits are they?  The most significant or the least significant?
Also, thinking about this further and from the standpoint of the end user writing the program - how will the user construct this string programatically in the test program?  I don't think it will be simply something like 
  exclude_any = "01010" 
Can we see an example?  Won't it be easy for the end user to create a 32 bit integer ?

Reply:
I tried to look this up in little more detail. The summary is there are two ways here : 
1) represent as hex string ( netconf yang model ) , 
2) global config where each bit ( or a subset ) in the 32 bit positions are  first assigned a string colour ( e.g. blue  = bit1 , red = bit2 , green=bit3 etc. ) and then the affinities are themselves represented as an 'array/list' of colours . [ across all routers / links / IGP and RSVP-TE tunnels ] 
We have always followed the 1st method in earlier configuration models. 
I have not seen any implementation that configures this as a normal integer ( which is much tougher to read from bits set point of view ) . 
Details:
1) Method 1: IETF working group ratified: 
http://www.netconfcentral.org/modules/ietf-te-types#admin-group.79
type yang:hex-string {
        length "1..11";
      }
      description
        "Administrative group / resource class / color representation in 'hex-string' type. The most significant byte in the hex-string is the farthest
to the left in the byte sequence.  Leading zero bytes in the configured value may be omitted for brevity."
[Apratim] We can of course add further clarification regarding MSB and omission of leading zeroes which seemed to be your one of your concerns. 
This IETF modeling of affinities closely match the proposed model [ exception : They use ':' between every 2 bytes , making max length as 11 instead of 8 for us , I feel this is not something very significant -> needs a new 4 byte MAC style which we dont have in OTG modeling options and does not seem to be a critical need as well] 

2) Method 2: Openconfig style: 
https://github.com/openconfig/public/blob/master/release/models/mpls/openconfig-mpls-te.yang
 leaf-list exclude-group {
      type leafref {
        path "../../../../../../../../../../te-global-attributes" +
          "/mpls-admin-groups/admin-group/admin-group-name";
      }
      description
        "list of references to named admin-groups to exclude in
        path calculation.";
    }
This follows the method of giving a name to each bit and then using array of such globally defined names to set the include/exclude groups for affinities. 
We can follow one of the two methods. Method1 seems easier to configure and implement compared to method 2. Then OTG should provide ways to define something like a dynamic global enum and allow
use of this enum for configuring protocol specific fields.
I have not changed anything yet in description or model . Based on your comments, I can add further clarification regarding MSB and leading zeros behaviour similar to IETF spec in all 6 such fields ( here and in Session Attributes) OR open up further discussion on adding new options to have global dynamic enum sort of element which can be re-used in protocol specific config with proper validation support from OTG/snappi layers .

Reply:
Thanks for the research.  Interesting & somewhat surprising (for me) findings.
I do agree that the second option, a separate enum for each bit, is a non-starter.
I also agree that if we go with the first option, then not having a ':' as a byte separator is fine.
I am still not clear how for the test writer constructing a string is easier.  Given that strings seem to be widely accepted here, it should be possible to find examples of how such strings are dynamically constructed (not just a static assignment)?  I'm really curious.

Reply:
Clarification on MSB and leading zeroes is added.  Most examples are CLI examples where the input is directly provided as hex bytes or colours.  Example go program snippets could possibly  look like this: 

type color int
const (
	red    color = 1 // bit 1 of 32
	blue         = 2
	green        = 4
	yellow       = 8   // bit 4 of 32
	purple       = 128 // bit 8 of 32
)
func xxx() {
	// Other stuff...
	include_any_as_hex_string := fmt.Sprintf("%X", yellow | blue | green | purple) 
	fmt.Println("Affinity : 0x", include_any_as_hex_string )
	// Other stuff...
}
$Affinity : 0x 8E

Reply:
Hmm..  The whole fmt.Sprintf() seems unnecessary and it does seem that simply marking it as an integer is a better option.  To me it seems more straight forward to do :
include_any_as_hex = yellow | blue | green | purple

And also, we could provide 'yellow', 'blue' as enums thru snappi/gosnappi.

@winstonliu1111 @ashutshkumr what do you guys think?

Reply:
Couple of notes : 
1) Note that the enums are user-defined i.e. which bits correspond to which 'colour' names is left to user.  It is not standardized. Different organizations can have different names for bits and they could something totally different  in different test programs/configs e.g. reduced_jitter = 0x1 , high_throughput = 0x2 etc. 
2)  It is easier to see the bits as hex when creating the config via json or proto text and not from script or when converting a config to json / pbtext and viewing the JSON output. e.g. 0x8E vs 142.
Probably that is the reason none of the existing specifications / CLI takes the bit inputs as integer.

Reply:
Agreed that hex notation is easier, hence my suggestion that it be modeled as an integer with hex input.  
@ashutshkumr @winstonliu1111 your thoughts on this?

Reply:
> 
@ankur-sheth isn't this already addressed?  I see "format: hex" added already.

Reply:
@winstonliu1111 Yes, it is "hex" format, but the type is a string (eg. "0x32").  See Apratim's example above of how the string would have to be constructed by the end user.